lib/dry/equalizer.rb

```diff
@@ -74,8 +75,16 @@ def define_cmp_method
     # @api private
     def define_hash_method
       keys = @keys
-      define_method(:hash) do | |
-        keys.map(&method(:send)).push(self.class).hash
+      immutable = @immutable
+
+      if immutable
+        define_method(:hash) do | |
+          @hash ||= keys.map(&method(:send)).push(self.class).hash
+        end
+      else
+        define_method(:hash) do
+          keys.map(&method(:send)).push(self.class).hash
```

Comment:
What do you think about doing something like this to avoid repetition?

```ruby
def define_hash_method
  keys = @keys
  immutable = @immutable
  define_method(:hash) do | |
    return @hash if immutable && defined?(@hash)
    @hash = keys.map(&method(:send)).push(self.class).hash
  end
end
```